### PR TITLE
feat: New module `avm/res/document-db/database-account`

### PR DESCRIPTION
**Pull request by krbar**
## Description

New Module `avm/res/document-db/database-account`, migrated from CARML.

## Adding a new module

- [x] A proposal has been submitted and approved.
- [ ] I have included "Closes #{module_proposal_issue_number}" in the PR description.
- [ ] I have run `brm validate` locally to verify the module files.
- [x] I have run deployment tests locally to ensure the module is deployable.

## Pipeline references

| Pipeline |
| - |
| [avm.res.document-db.database-account](https://github.com/krbar/bicep-registry-modules/actions/workflows/avm.res.document-db.database-account.yml) |

### CODE REVIEW

File: avm/res/document-db/database-account/gremlin-database/main.bicep [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+param databaseAccountName string
+
+@description('Optional. Array of graphs to deploy in the Gremlin database.')
+param graphs array = []
+
+@description('Optional. Represents maximum throughput, the resource can scale up to. Cannot be set together with `throughput`. If `throughput` is set to something else than -1, this autoscale setting is ignored.')
+param maxThroughput int = 4000
+
+@description('Optional. Request Units per second (for example 10000). Cannot be set together with `maxThroughput`.')
+param throughput int = -1
```
Comment (AlexanderSehr) [inline avm/res/document-db/database-account/gremlin-database/main.bicep:21]: Please make nullable

File: avm/res/document-db/database-account/mongodb-database/collection/main.bicep [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    }
+    resource: {
+      id: name
+      indexes: indexes
+      shardKey: shardKey
+    }
+  }
+}
+
+@description('The name of the mongodb database.')
```
Comment (AlexanderSehr) [inline avm/res/document-db/database-account/mongodb-database/collection/main.bicep:46]: This and the other descriptions are incorrectly referring to. MongoDB

File: avm/res/document-db/database-account/mongodb-database/main.bicep [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    }
+    options: contains(databaseAccount.properties.capabilities, { name: 'EnableServerless' }) ? null : {
+      throughput: throughput
+    }
+  }
+}
+
+module mongodbDatabase_collections 'collection/main.bicep' = [for collection in collections: {
+  name: '${uniqueString(deployment().name, mongodbDatabase.name)}-collection-${collection.name}'
+  params: {
```
Comment (AlexanderSehr) [inline avm/res/document-db/database-account/mongodb-database/main.bicep:40]: I may be mistaken, but I think this is not passing through all parameters

File: avm/res/document-db/database-account/sql-database/container/main.bicep [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+@description('Optional. Indicates how long data should be retained in the analytical store, for a container. Analytical store is enabled when ATTL is set with a value other than 0. If the value is set to -1, the analytical store retains all historical data, irrespective of the retention of the data in the transactional store.')
+param analyticalStorageTtl int = 0
+
+@description('Optional. The conflict resolution policy for the container. Conflicts and conflict resolution policies are applicable if the Azure Cosmos DB account is configured with multiple write regions.')
+param conflictResolutionPolicy object = {}
+
+@maxValue(2147483647)
+@minValue(-1)
+@description('Optional. Default time to live (in seconds). With Time to Live or TTL, Azure Cosmos DB provides the ability to delete items automatically from a container after a certain time period. If the value is set to "-1", it is equal to infinity, and items dont expire by default.')
+param defaultTtl int = -1
```
Comment (AlexanderSehr) [inline avm/res/document-db/database-account/sql-database/container/main.bicep:23]: Please make nullable

File: avm/res/document-db/database-account/sql-database/container/main.bicep [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+@description('Optional. Request Units per second. Will be set to null if autoscaleSettingsMaxThroughput is used.')
+param throughput int = 400
+
+@maxValue(1000000)
+@description('Optional. Specifies the Autoscale settings and represents maximum throughput, the resource can scale up to. The autoscale throughput should have valid throughput values between 1000 and 1000000 inclusive in increments of 1000. If value is set to -1, then the property will be set to null and autoscale will be disabled.')
+param autoscaleSettingsMaxThroughput int = -1
```
Comment (AlexanderSehr) [inline avm/res/document-db/database-account/sql-database/container/main.bicep:30]: Please make nullable

File: avm/res/document-db/database-account/sql-database/main.bicep [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+param name string
+
+@description('Optional. Array of containers to deploy in the SQL database.')
+param containers array = []
+
+@description('Optional. Request units per second. Will be set to null if autoscaleSettingsMaxThroughput is used.')
+param throughput int = 400
+
+@description('Optional. Specifies the Autoscale settings and represents maximum throughput, the resource can scale up to.  The autoscale throughput should have valid throughput values between 1000 and 1000000 inclusive in increments of 1000. If value is set to -1, then the property will be set to null and autoscale will be disabled.')
+param autoscaleSettingsMaxThroughput int = -1
```
Comment (AlexanderSehr) [inline avm/res/document-db/database-account/sql-database/main.bicep:18]: Please make nullable

File: avm/res/document-db/database-account/tests/e2e/mongodb/main.test.bicep [hunk 7]
```diff
@@ -0,0 +1,304 @@
...
+    eventHubNamespaceName: 'dep-${namePrefix}-evhns-${serviceShort}'
+    location: location
+  }
+}
+
+// ============== //
+// Test Execution //
+// ============== //
+
+module testDeployment '../../../main.bicep' = {
```
Comment (AlexanderSehr) [inline avm/res/document-db/database-account/tests/e2e/mongodb/main.test.bicep:62]: For this and the other tests I assume the idempotency failed? Anything we can do here?

**Review by AlexanderSehr — CHANGES_REQUESTED** [review]

File: avm/res/document-db/database-account/gremlin-database/main.bicep (hunk 1 above)
Comment (krbar) [reply avm/res/document-db/database-account/gremlin-database/main.bicep]: done: https://github.com/Azure/bicep-registry-modules/pull/667/commits/49d95a9ca8c6ed3c67a9715e9777da14770ca5b9

File: avm/res/document-db/database-account/mongodb-database/collection/main.bicep (hunk 2 above)
Comment (krbar) [reply avm/res/document-db/database-account/mongodb-database/collection/main.bicep]: Fixed: https://github.com/Azure/bicep-registry-modules/pull/667/commits/0e85651de73c0c7053a3cd8e6bd152ba5ab374d1

File: avm/res/document-db/database-account/mongodb-database/main.bicep (hunk 3 above)
Comment (krbar) [reply avm/res/document-db/database-account/mongodb-database/main.bicep:40]: You were right, fixed: https://github.com/Azure/bicep-registry-modules/pull/667/commits/27f7c366f2abaa7dbc7d053c6b2a9c1fbc7b6f45

File: avm/res/document-db/database-account/main.bicep [hunk 8]
```diff
@@ -0,0 +1,504 @@
...
+    databaseAccountName: databaseAccount.name
+    name: mongodbDatabase.name
+    collections: contains(mongodbDatabase, 'collections') ? mongodbDatabase.collections : []
+  }
+}]
+
+module databaseAccount_gremlinDatabases 'gremlin-database/main.bicep' = [for gremlinDatabase in gremlinDatabases: {
+  name: '${uniqueString(deployment().name, location)}-gremlin-${gremlinDatabase.name}'
+  params: {
+    databaseAccountName: databaseAccount.name
```
Comment (AlexanderSehr) [inline avm/res/document-db/database-account/main.bicep:312]: Could it be that for this and other child resources the majority of parameters isn't passed down?

**Review by AlexanderSehr — CHANGES_REQUESTED** [review]

**Review by skydiver1980 — APPROVED** [review]

**Review by skydiver1980 — APPROVED** [review]

File: avm/res/document-db/database-account/main.bicep (hunk 8 above)
Comment (krbar) [reply avm/res/document-db/database-account/main.bicep:314]: good catch, you are right. Fixed in ead490de661e1e5653f8bbe906c28d2a7c0935af

File: avm/res/document-db/database-account/sql-database/main.bicep (hunk 6 above)
Comment (krbar) [reply avm/res/document-db/database-account/sql-database/main.bicep]: implemented in b9e0bc4e17a4442fab0428d88e20b7361fa3a305

File: avm/res/document-db/database-account/sql-database/container/main.bicep (hunk 5 above)
Comment (krbar) [reply avm/res/document-db/database-account/sql-database/container/main.bicep]: Implemented in b9e0bc4e17a4442fab0428d88e20b7361fa3a305

File: avm/res/document-db/database-account/sql-database/container/main.bicep (hunk 4 above)
Comment (krbar) [reply avm/res/document-db/database-account/sql-database/container/main.bicep:23]: As discussed on Teams, the "-1" here is not just a workaround, but a documented value, so I suggest to keep it.
See: https://learn.microsoft.com/en-us/azure/cosmos-db/nosql/time-to-live

**Review by AlexanderSehr — CHANGES_REQUESTED** [review]

**Review by AlexanderSehr — APPROVED** [review]

**Pull request merged**